torch/csrc/autograd/engine.cpp

```diff
@@ -587,6 +587,9 @@ void Engine::evaluate_function(
             inputs[capture.input_idx_];
       }
     }
+    for (auto& hook : fn_info.hooks_) {
+      (*hook)(inputs.toVariables());
+    }
```

Comment:
If this is a grad capture post hook, these hooks should be executed only if `fn_info.needed_` is false?

Reply:
I want to make the hook execution orthogonal to `needed_`. The callsite has `neede_` info and can decide whether to install the hook accordingly if necessary.
Otherwise, a callsite may get surprised if the hook of a not-needed node is not executed in the future.